<input type=color alpha> swatch does not have a background that reflects the transparency

annevk opened the pull request:
#### 3d975165870fdedf358571e31742bfa93eab1fd3
<pre>
&lt;input type=color alpha&gt; swatch does not have a background that reflects the transparency
<a href="https://bugs.webkit.org/show_bug.cgi?id=286212">https://bugs.webkit.org/show_bug.cgi?id=286212</a>
<a href="https://rdar.apple.com/143188497">rdar://143188497</a>

Reviewed by Aditya Keerthi.

When a color with alpha channel was selected this would result in the
background of the parent element becoming visible. To prevent this we
perform a source-over color blending operation with opaque white. This
matches how the control behaves on iOS and seems like a reasonable
platform-neutral default.

For macOS we want to render a diagonally divided rectangle with one of
the triangles using opaque black as input for the blending operation
and the other opaque white to match the platform color well.

* LayoutTests/fast/forms/color/color-input-swatch-expected.txt: Added.
* LayoutTests/fast/forms/color/color-input-swatch.html: Added.
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt: Added.
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::updateColorSwatch):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::setColorWellSwatchBackground):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::setColorWellSwatchBackground):

Canonical link: <a href="https://commits.webkit.org/289314@main">https://commits.webkit.org/289314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/887aa6d5a3be8a3a9ab284f25a5568ad655c975e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86361 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5913 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40698 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91274 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37164 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66876 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24671 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78232 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4493 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32549 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75034 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93099 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75669 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74850 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19106 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17502 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13444 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13589 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18878 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13342 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16785 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15126 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (approved):
You could probably write a layout test for this using `internals.shadowRoot` to access the swatch element, and checking computed style.

annevk committed:
<input type=color alpha> swatch does not have a background that reflects the transparency
https://bugs.webkit.org/show_bug.cgi?id=286212
rdar://143188497

Reviewed by Aditya Keerthi.

When a color with alpha channel was selected this would result in the
background of the parent element becoming visible. To prevent this we
perform a source-over color blending operation with opaque white. This
matches how the control behaves on iOS and seems like a reasonable
platform-neutral default.

For macOS we want to render a diagonally divided rectangle with one of
the triangles using opaque black as input for the blending operation
and the other opaque white to match the platform color well.

* LayoutTests/fast/forms/color/color-input-swatch-expected.txt: Added.
* LayoutTests/fast/forms/color/color-input-swatch.html: Added.
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt: Added.
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::updateColorSwatch):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::setColorWellSwatchBackground):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::setColorWellSwatchBackground):

Canonical link: https://commits.webkit.org/289314@main
